module/zfs/qat_compress.c

```diff
@@ -534,7 +552,43 @@ qat_compress(qat_compress_dir_t dir, char *src, int src_len,
 	return (ret);
 }
 
-module_param(zfs_qat_compress_disable, int, 0644);
-MODULE_PARM_DESC(zfs_qat_compress_disable, "Disable QAT compression");
+static int
+#ifdef module_param_cb
+param_set_qat_compress(const char *val, const struct kernel_param *kp)
+#else
+param_set_qat_compress(const char *val, struct kernel_param *kp)
+#endif
+{
+	int ret;
+	int *pvalue = kp->arg;
+	ret = param_set_int(val, kp);
+	if (ret)
+		return (ret);
+	/*
+	 * zfs_qat_compress_disable = 0: enable qat compress
+	 * try to initialize qat instance if it has not been done
+	 */
+	if (*pvalue == 0 && !qat_dc_init_done) {
+		ret = qat_dc_init();
+		if (ret != 0) {
+			zfs_qat_compress_disable = 1;
+			return (ret);
+		}
+	}
+	return (ret);
+}
+
+#ifdef module_param_cb
+static struct kernel_param_ops qat_compress_param_ops = {
+	.set = param_set_qat_compress,
+	.get = param_get_int,
+};
+module_param_cb(zfs_qat_compress_disable, &qat_compress_param_ops,
+	&zfs_qat_compress_disable, 0644);
+#else
+module_param_call(zfs_qat_compress_disable, param_set_qat_compress,
+    param_get_int, 0644);
+#endif
```

Comment:
Compatibly code was added in `include/linux/mod_compat.h` to help hide some of this conditional ugliness.  Can you structure this in a similar way to the `zfs_multihost_interval` module option, here:

https://github.com/zfsonlinux/zfs/blob/master/module/zfs/mmp.c#L722L745

You'd need to update the other two module parameters in the same way.